### DISCUSSION TITLE
Play midi via AlphaSynth does not work in node environment.

### DISCUSSION THREAD

**Discussion by Barba828** [CoderLine/alphaTab#1606] [Q&A] [answered]
Running in Electron, I tried several api, but none worked. There must be something wrong with the way I use it.
**There are some error:**
<img width="289" alt="image" src="https://github.com/user-attachments/assets/301200bd-3cf9-44f8-baff-33df031e246a">

**There are some refer:**
refer_1:  https://alphatab.net/docs/guides/lowlevel-apis
refer_2: https://github.com/CoderLine/alphaTab/discussions/895

**code:**
```js
	playMidi = async () => {
		const soundFontResponse = await fetch(
			"https://barba828.github.io/buitar-editor/soundfont/sonivox.sf2"
		);
		const midiFile = new alphaTab.midi.MidiFile();
		const soundFont = new Uint8Array(await soundFontResponse.arrayBuffer());

		// test_1
		const player = new alphaTab.synth.AlphaSynth(
			new alphaTab.synth.AlphaSynthAudioWorkletOutput(this.alphaTabSettings),
			99999999
		);

		// test_2
		// const player = new alphaTab.synth.AlphaSynthWebWorkerApi(
		// 	new  alphaTab.synth.AlphaSynthAudioWorkletOutput(this.alphaTabSettings),
		// 	this.alphaTabSettings
		// );

		// test_3
		// const player = new alphaTab.synth.AlphaSynthWebWorkerApi(
		// 	new alphaTab.synth.AlphaSynthScriptProcessorOutput(),
		// 	this.alphaTabSettings
		// );

		player.loadSoundFont(soundFont, false);
		player.loadMidiFile(midiFile);
		player.play();
	};
```

**Comment by Danielku15** [comment, marked as answer]
I think we might be mixing things here a bit. 

1. Audio Synthesis in Node.

The `AlphaSynthAudioWorkletOutput` is not supported on Node. As the name implies: This "output" is using Audio Worklets which are a Browser API unavailable in Node. We currently have no output which would work on node.js. As node.js doesn't ship a platform API itself for audio, some third party library would have to be used. 

If you just want to collect the raw audio samples yourself a implementation of an output can look like this: https://github.com/CoderLine/alphaTab/blob/develop/test/audio/TestOutput.ts

* https://github.com/ircam-ismm/node-web-audio-api

1. Electron and Web Audio. 

Electron is generally using the Chromium render engine. Assuming you use alphaTab in the "browser" part of Electron, you can follow most parts of using alphaTab as from a web application. To load the SoundFont and input files you would use the electron bits to either get blobs (Blob,ArrayBuffer,Uint8Array) to pass to alphaTab. [^1]

There might be some projects providing Web Audio in Node.js scopes. The challenge is that node.js is single threaded and alphaTab in the browser mode launches background workers to handle the multi-threading aspects. These platform aspects are not available in node.js either.

If you use electron, my recommendation would be to stay on the browser level for alphaTab display and playback. You might use alphaTab also on node.js level for some organizational bits (e.g. reading song details from files for a media library). 


[^1]: https://medium.com/@Jeff_Duke_io/working-with-html5-audio-in-electron-645b2d2202bd
